File: rust/src/move_generator.rs
use crate::types::*;
use crate::board::Board;

pub struct MoveGenerator;

impl MoveGenerator {
    pub fn new() -> Self {
        Self
    }

    pub fn generate_moves(&self, board: &Board, color: Color) -> Vec<Move> {
        let mut moves = Vec::new();
        
        for square in 0..64 {
            if let Some(piece) = board.get_piece(square) {
                if piece.color == color {
                    moves.extend(self.generate_piece_moves(board, square, piece));
                }
            }
        }
        
        moves
    }

    fn generate_piece_moves(&self, board: &Board, from: Square, piece: Piece) -> Vec<Move> {
        match piece.piece_type {
            PieceType::Pawn => self.generate_pawn_moves(board, from, piece.color),
            PieceType::Knight => self.generate_knight_moves(board, from, piece.color),
            PieceType::Bishop => self.generate_bishop_moves(board, from, piece.color),
            PieceType::Rook => self.generate_rook_moves(board, from, piece.color),
            PieceType::Queen => self.generate_queen_moves(board, from, piece.color),
            PieceType::King => self.generate_king_moves(board, from, piece.color),
        }
    }

    fn generate_pawn_moves(&self, board: &Board, from: Square, color: Color) -> Vec<Move> {
        let mut moves = Vec::new();
        let direction = if color == Color::White { 8 } else { -8i32 } as i32;
        let start_rank = if color == Color::White { 1 } else { 6 };
        let promotion_rank = if color == Color::White { 7 } else { 0 };
        
        let from_i32 = from as i32;
        let rank = from / 8;
        let file = from % 8;

        // One square forward
        let one_forward = from_i32 + direction;
        if self.is_valid_square(one_forward) && board.get_piece(one_forward as usize).is_none() {
            let to = one_forward as usize;
            if to / 8 == promotion_rank {
                // Promotion moves
                for promotion_piece in [PieceType::Queen, PieceType::Rook, PieceType::Bishop, PieceType::Knight] {
                    moves.push(Move::new(from, to, PieceType::Pawn).with_promotion(promotion_piece));
                }
            } else {
                moves.push(Move::new(from, to, PieceType::Pawn));
            }

            // Two squares forward from starting position
            if rank == start_rank {
                let two_forward = from_i32 + 2 * direction;
                if self.is_valid_square(two_forward) && board.get_piece(two_forward as usize).is_none() {
                    moves.push(Move::new(from, two_forward as usize, PieceType::Pawn));
                }
            }
        }

        // Captures
        for &offset in &[direction - 1, direction + 1] {
            let to = from_i32 + offset;
            let to_file = (to % 8) as usize;
            
            if self.is_valid_square(to) && (to_file as i32 - file as i32).abs() == 1 {
                let to_square = to as usize;
                if let Some(target_piece) = board.get_piece(to_square) {
                    if target_piece.color != color {
                        if to_square / 8 == promotion_rank {
                            // Promotion captures
                            for promotion_piece in [PieceType::Queen, PieceType::Rook, PieceType::Bishop, PieceType::Knight] {
                                moves.push(Move::new(from, to_square, PieceType::Pawn)
                                    .with_capture(target_piece.piece_type)
                                    .with_promotion(promotion_piece));
                            }
                        } else {
                            moves.push(Move::new(from, to_square, PieceType::Pawn)
                                .with_capture(target_piece.piece_type));
                        }
                    }
                }
            }
        }

        // En passant
        if let Some(en_passant_target) = board.get_en_passant_target() {
            let expected_rank = if color == Color::White { 4 } else { 3 };
            if rank == expected_rank {
                for &offset in &[direction - 1, direction + 1] {
                    let to = from_i32 + offset;
                    if self.is_valid_square(to) && to as usize == en_passant_target {
                        moves.push(Move::new(from, to as usize, PieceType::Pawn)
                            .with_capture(PieceType::Pawn)
                            .with_en_passant());
                    }
                }
            }
        }

        moves
    }

    fn generate_knight_moves(&self, board: &Board, from: Square, color: Color) -> Vec<Move> {
        let mut moves = Vec::new();
        let offsets = [-17, -15, -10, -6, 6, 10, 15, 17];
        let from_i32 = from as i32;
        let file = from % 8;

        for offset in offsets {
            let to = from_i32 + offset;
            let to_file = (to % 8) as usize;

            if self.is_valid_square(to) && (to_file as i32 - file as i32).abs() <= 2 {
                let to_square = to as usize;
                match board.get_piece(to_square) {
                    None => moves.push(Move::new(from, to_square, PieceType::Knight)),
                    Some(piece) if piece.color != color => {
                        moves.push(Move::new(from, to_square, PieceType::Knight)
                            .with_capture(piece.piece_type));
                    },
                    _ => {}
                }
            }
        }

        moves
    }

    fn generate_bishop_moves(&self, board: &Board, from: Square, color: Color) -> Vec<Move> {
        self.generate_sliding_moves(board, from, color, &[-9, -7, 7, 9], PieceType::Bishop)
    }

    fn generate_rook_moves(&self, board: &Board, from: Square, color: Color) -> Vec<Move> {
        self.generate_sliding_moves(board, from, color, &[-8, -1, 1, 8], PieceType::Rook)
    }

    fn generate_queen_moves(&self, board: &Board, from: Square, color: Color) -> Vec<Move> {
        self.generate_sliding_moves(board, from, color, &[-9, -8, -7, -1, 1, 7, 8, 9], PieceType::Queen)
    }

    fn generate_king_moves(&self, board: &Board, from: Square, color: Color) -> Vec<Move> {
        let mut moves = Vec::new();
        let offsets = [-9, -8, -7, -1, 1, 7, 8, 9];
        let from_i32 = from as i32;
        let file = from % 8;

        for offset in offsets {
            let to = from_i32 + offset;
            let to_file = (to % 8) as usize;

            if self.is_valid_square(to) && (to_file as i32 - file as i32).abs() <= 1 {
                let to_square = to as usize;
                match board.get_piece(to_square) {
                    None => moves.push(Move::new(from, to_square, PieceType::King)),
                    Some(piece) if piece.color != color => {
                        moves.push(Move::new(from, to_square, PieceType::King)
                            .with_capture(piece.piece_type));
                    },
                    _ => {}
                }
            }
        }

        // Castling
        let rights = board.get_castling_rights();
        if color == Color::White && from == 4 {
            // White kingside
            if rights.white_kingside && 
               board.get_piece(5).is_none() && 
               board.get_piece(6).is_none() &&
               board.get_piece(7).map_or(false, |p| p.piece_type == PieceType::Rook && p.color == Color::White) {
                if !self.is_square_attacked(board, 4, Color::Black) &&
                   !self.is_square_attacked(board, 5, Color::Black) &&
                   !self.is_square_attacked(board, 6, Color::Black) {
                    moves.push(Move::new(4, 6, PieceType::King).with_castling());
                }
            }
            // White queenside
            if rights.white_queenside &&
               board.get_piece(3).is_none() &&
               board.get_piece(2).is_none() &&
               board.get_piece(1).is_none() &&
               board.get_piece(0).map_or(false, |p| p.piece_type == PieceType::Rook && p.color == Color::White) {
                if !self.is_square_attacked(board, 4, Color::Black) &&
                   !self.is_square_attacked(board, 3, Color::Black) &&
                   !self.is_square_attacked(board, 2, Color::Black) {
                    moves.push(Move::new(4, 2, PieceType::King).with_castling());
                }
            }
        } else if color == Color::Black && from == 60 {
            // Black kingside
            if rights.black_kingside &&
               board.get_piece(61).is_none() &&
               board.get_piece(62).is_none() &&
               board.get_piece(63).map_or(false, |p| p.piece_type == PieceType::Rook && p.color == Color::Black) {
                if !self.is_square_attacked(board, 60, Color::White) &&
                   !self.is_square_attacked(board, 61, Color::White) &&
                   !self.is_square_attacked(board, 62, Color::White) {
                    moves.push(Move::new(60, 62, PieceType::King).with_castling());
                }
            }
            // Black queenside
            if rights.black_queenside &&
               board.get_piece(59).is_none() &&
               board.get_piece(58).is_none() &&
               board.get_piece(57).is_none() &&
               board.get_piece(56).map_or(false, |p| p.piece_type == PieceType::Rook && p.color == Color::Black) {
                if !self.is_square_attacked(board, 60, Color::White) &&
                   !self.is_square_attacked(board, 59, Color::White) &&
                   !self.is_square_attacked(board, 58, Color::White) {
                    moves.push(Move::new(60, 58, PieceType::King).with_castling());
                }
            }
        }

        moves
    }

    fn generate_sliding_moves(&self, board: &Board, from: Square, color: Color, directions: &[i32], piece_type: PieceType) -> Vec<Move> {
        let mut moves = Vec::new();
        let from_i32 = from as i32;

        for &direction in directions {
            let mut to = from_i32 + direction;
            let mut prev_file = (from % 8) as i32;

            while self.is_valid_square(to) {
                let to_file = to % 8;
                
                // Check for wrapping (especially important for horizontal moves)
                if direction == -1 || direction == 1 {
                    if (to_file - prev_file).abs() != 1 {
                        break;
                    }
                }

                let to_square = to as usize;
                match board.get_piece(to_square) {
                    None => {
                        moves.push(Move::new(from, to_square, piece_type));
                    },
                    Some(piece) => {
                        if piece.color != color {
                            moves.push(Move::new(from, to_square, piece_type)
                                .with_capture(piece.piece_type));
                        }
                        break;
                    }
                }

                prev_file = to_file;
                to += direction;
            }
        }

        moves
    }

    pub fn is_square_attacked(&self, board: &Board, square: Square, by_color: Color) -> bool {
        for from in 0..64 {
            if let Some(piece) = board.get_piece(from) {
                if piece.color == by_color {
                    let moves = self.generate_piece_moves(board, from, piece);
                    if moves.iter().any(|m| m.to == square) {
                        return true;
                    }
                }
            }
        }
        false
    }

    pub fn is_in_check(&self, board: &Board, color: Color) -> bool {
        for square in 0..64 {
            if let Some(piece) = board.get_piece(square) {
                if piece.color == color && piece.piece_type == PieceType::King {
                    return self.is_square_attacked(board, square, color.opposite());
                }
            }
        }
        false
    }

    pub fn get_legal_moves(&self, board: &Board, color: Color) -> Vec<Move> {
        let moves = self.generate_moves(board, color);
        let mut legal_moves = Vec::new();

        for chess_move in moves {
            let mut board_copy = board.get_state().clone();
            let mut test_board = Board::new();
            test_board.set_state(board_copy);
            
            test_board.make_move(&chess_move);
            
            if !self.is_in_check(&test_board, color) {
                legal_moves.push(chess_move);
            }
        }

        legal_moves
    }

    pub fn is_checkmate(&self, board: &Board, color: Color) -> bool {
        self.is_in_check(board, color) && self.get_legal_moves(board, color).is_empty()
    }

    pub fn is_stalemate(&self, board: &Board, color: Color) -> bool {
        !self.is_in_check(board, color) && self.get_legal_moves(board, color).is_empty()
    }

    fn is_valid_square(&self, square: i32) -> bool {
        square >= 0 && square < 64
    }
}

File: zig/src/move_generator.zig
const std = @import("std");
const board = @import("board.zig");

pub const MoveGenerator = struct {
    board_ref: *board.Board,

    pub fn init(board_ref: *board.Board) MoveGenerator {
        return MoveGenerator{
            .board_ref = board_ref,
        };
    }

    pub fn generateLegalMoves(self: *MoveGenerator, allocator: std.mem.Allocator) !std.ArrayList(board.Move) {
        var moves = std.ArrayList(board.Move).init(allocator);
        
        const current_color = if (self.board_ref.white_to_move) board.PieceColor.White else board.PieceColor.Black;
        
        for (self.board_ref.squares, 0..) |piece, i| {
            if (piece) |p| {
                if (p.color == current_color) {
                    try self.generatePieceMoves(@intCast(i), p, &moves);
                }
            }
        }
        
        return moves;
    }

    fn generatePieceMoves(self: *MoveGenerator, from: u8, piece: board.Piece, moves: *std.ArrayList(board.Move)) !void {
        switch (piece.piece_type) {
            .Pawn => try self.generatePawnMoves(from, piece.color, moves),
            .Knight => try self.generateKnightMoves(from, piece.color, moves),
            .Bishop => try self.generateBishopMoves(from, piece.color, moves),
            .Rook => try self.generateRookMoves(from, piece.color, moves),
            .Queen => try self.generateQueenMoves(from, piece.color, moves),
            .King => try self.generateKingMoves(from, piece.color, moves),
        }
    }

    fn generatePawnMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, moves: *std.ArrayList(board.Move)) !void {
        const from_rank = from / 8;
        const from_file = from % 8;
        const direction: i8 = if (color == .White) 1 else -1;
        const starting_rank = if (color == .White) 1 else 6;
        const promotion_rank = if (color == .White) 7 else 0;

        // Forward moves
        const one_forward = @as(i8, @intCast(from)) + direction * 8;
        if (one_forward >= 0 and one_forward < 64) {
            const to_square = @as(u8, @intCast(one_forward));
            if (self.board_ref.squares[to_square] == null) {
                if (to_square / 8 == promotion_rank) {
                    // Promotion moves
                    try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Queen });
                    try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Rook });
                    try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Bishop });
                    try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Knight });
                } else {
                    try moves.append(board.Move{ .from = from, .to = to_square });
                }

                // Two squares forward from starting position
                if (from_rank == starting_rank) {
                    const two_forward = @as(i8, @intCast(from)) + direction * 16;
                    if (two_forward >= 0 and two_forward < 64) {
                        const to_square_two = @as(u8, @intCast(two_forward));
                        if (self.board_ref.squares[to_square_two] == null) {
                            try moves.append(board.Move{ .from = from, .to = to_square_two });
                        }
                    }
                }
            }
        }

        // Captures
        const capture_moves = [_]i8{ direction * 8 - 1, direction * 8 + 1 };
        for (capture_moves) |move_offset| {
            const target = @as(i8, @intCast(from)) + move_offset;
            if (target >= 0 and target < 64) {
                const to_square = @as(u8, @intCast(target));
                const to_file = to_square % 8;
                
                // Check file bounds for diagonal captures
                if ((move_offset == direction * 8 - 1 and from_file > 0) or
                    (move_offset == direction * 8 + 1 and from_file < 7)) {
                    
                    if (self.board_ref.squares[to_square]) |target_piece| {
                        if (target_piece.color != color) {
                            if (to_square / 8 == promotion_rank) {
                                // Promotion captures
                                try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Queen });
                                try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Rook });
                                try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Bishop });
                                try moves.append(board.Move{ .from = from, .to = to_square, .promotion_piece = .Knight });
                            } else {
                                try moves.append(board.Move{ .from = from, .to = to_square });
                            }
                        }
                    } else if (self.board_ref.en_passant_target == to_square) {
                        // En passant capture
                        try moves.append(board.Move{ .from = from, .to = to_square, .en_passant = true });
                    }
                }
            }
        }
    }

    fn generateKnightMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, moves: *std.ArrayList(board.Move)) !void {
        const from_rank = @as(i8, @intCast(from / 8));
        const from_file = @as(i8, @intCast(from % 8));
        
        const knight_moves = [_][2]i8{
            .{ 2, 1 }, .{ 2, -1 }, .{ -2, 1 }, .{ -2, -1 },
            .{ 1, 2 }, .{ 1, -2 }, .{ -1, 2 }, .{ -1, -2 }
        };

        for (knight_moves) |move| {
            const to_rank = from_rank + move[0];
            const to_file = from_file + move[1];

            if (to_rank >= 0 and to_rank < 8 and to_file >= 0 and to_file < 8) {
                const to_square = @as(u8, @intCast(to_rank)) * 8 + @as(u8, @intCast(to_file));
                
                if (self.board_ref.squares[to_square]) |target_piece| {
                    if (target_piece.color != color) {
                        try moves.append(board.Move{ .from = from, .to = to_square });
                    }
                } else {
                    try moves.append(board.Move{ .from = from, .to = to_square });
                }
            }
        }
    }

    fn generateBishopMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, moves: *std.ArrayList(board.Move)) !void {
        const directions = [_][2]i8{ .{ 1, 1 }, .{ 1, -1 }, .{ -1, 1 }, .{ -1, -1 } };
        try self.generateSlidingMoves(from, color, &directions, moves);
    }

    fn generateRookMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, moves: *std.ArrayList(board.Move)) !void {
        const directions = [_][2]i8{ .{ 1, 0 }, .{ -1, 0 }, .{ 0, 1 }, .{ 0, -1 } };
        try self.generateSlidingMoves(from, color, &directions, moves);
    }

    fn generateQueenMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, moves: *std.ArrayList(board.Move)) !void {
        const directions = [_][2]i8{
            .{ 1, 1 }, .{ 1, -1 }, .{ -1, 1 }, .{ -1, -1 },
            .{ 1, 0 }, .{ -1, 0 }, .{ 0, 1 }, .{ 0, -1 }
        };
        try self.generateSlidingMoves(from, color, &directions, moves);
    }

    fn generateSlidingMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, directions: []const [2]i8, moves: *std.ArrayList(board.Move)) !void {
        const from_rank = @as(i8, @intCast(from / 8));
        const from_file = @as(i8, @intCast(from % 8));

        for (directions) |direction| {
            var rank = from_rank + direction[0];
            var file = from_file + direction[1];

            while (rank >= 0 and rank < 8 and file >= 0 and file < 8) {
                const to_square = @as(u8, @intCast(rank)) * 8 + @as(u8, @intCast(file));

                if (self.board_ref.squares[to_square]) |target_piece| {
                    if (target_piece.color != color) {
                        try moves.append(board.Move{ .from = from, .to = to_square });
                    }
                    break; // Can't move past any piece
                } else {
                    try moves.append(board.Move{ .from = from, .to = to_square });
                }

                rank += direction[0];
                file += direction[1];
            }
        }
    }

    fn generateKingMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, moves: *std.ArrayList(board.Move)) !void {
        const from_rank = @as(i8, @intCast(from / 8));
        const from_file = @as(i8, @intCast(from % 8));

        const king_moves = [_][2]i8{
            .{ 1, 1 }, .{ 1, -1 }, .{ -1, 1 }, .{ -1, -1 },
            .{ 1, 0 }, .{ -1, 0 }, .{ 0, 1 }, .{ 0, -1 }
        };

        for (king_moves) |move| {
            const to_rank = from_rank + move[0];
            const to_file = from_file + move[1];

            if (to_rank >= 0 and to_rank < 8 and to_file >= 0 and to_file < 8) {
                const to_square = @as(u8, @intCast(to_rank)) * 8 + @as(u8, @intCast(to_file));
                
                if (self.board_ref.squares[to_square]) |target_piece| {
                    if (target_piece.color != color) {
                        try moves.append(board.Move{ .from = from, .to = to_square });
                    }
                } else {
                    try moves.append(board.Move{ .from = from, .to = to_square });
                }
            }
        }

        // Castling moves
        try self.generateCastlingMoves(from, color, moves);
    }

    fn generateCastlingMoves(self: *MoveGenerator, from: u8, color: board.PieceColor, moves: *std.ArrayList(board.Move)) !void {
        if (self.board_ref.isInCheck(color)) return; // Can't castle out of check

        if (color == .White and from == 4) {
            // White castling
            if (self.board_ref.castling_rights.white_king_side and
                self.board_ref.squares[5] == null and
                self.board_ref.squares[6] == null and
                self.board_ref.squares[7] != null and
                self.board_ref.squares[7].?.piece_type == .Rook) {
                
                // Check that king doesn't pass through check
                if (!self.wouldBeInCheck(4, 5, color) and !self.wouldBeInCheck(4, 6, color)) {
                    try moves.append(board.Move{ .from = from, .to = 6, .castle_king_side = true });
                }
            }

            if (self.board_ref.castling_rights.white_queen_side and
                self.board_ref.squares[3] == null and
                self.board_ref.squares[2] == null and
                self.board_ref.squares[1] == null and
                self.board_ref.squares[0] != null and
                self.board_ref.squares[0].?.piece_type == .Rook) {
                
                // Check that king doesn't pass through check
                if (!self.wouldBeInCheck(4, 3, color) and !self.wouldBeInCheck(4, 2, color)) {
                    try moves.append(board.Move{ .from = from, .to = 2, .castle_queen_side = true });
                }
            }
        } else if (color == .Black and from == 60) {
            // Black castling
            if (self.board_ref.castling_rights.black_king_side and
                self.board_ref.squares[61] == null and
                self.board_ref.squares[62] == null and
                self.board_ref.squares[63] != null and
                self.board_ref.squares[63].?.piece_type == .Rook) {
                
                // Check that king doesn't pass through check
                if (!self.wouldBeInCheck(60, 61, color) and !self.wouldBeInCheck(60, 62, color)) {
                    try moves.append(board.Move{ .from = from, .to = 62, .castle_king_side = true });
                }
            }

            if (self.board_ref.castling_rights.black_queen_side and
                self.board_ref.squares[59] == null and
                self.board_ref.squares[58] == null and
                self.board_ref.squares[57] == null and
                self.board_ref.squares[56] != null and
                self.board_ref.squares[56].?.piece_type == .Rook) {
                
                // Check that king doesn't pass through check
                if (!self.wouldBeInCheck(60, 59, color) and !self.wouldBeInCheck(60, 58, color)) {
                    try moves.append(board.Move{ .from = from, .to = 58, .castle_queen_side = true });
                }
            }
        }
    }

    fn wouldBeInCheck(self: *MoveGenerator, from: u8, to: u8, color: board.PieceColor) bool {
        // Temporarily make the move and check if in check
        const original_piece = self.board_ref.squares[from];
        const captured_piece = self.board_ref.squares[to];
        
        self.board_ref.squares[from] = null;
        self.board_ref.squares[to] = original_piece;
        
        const in_check = self.board_ref.isInCheck(color);
        
        // Restore original position
        self.board_ref.squares[from] = original_piece;
        self.board_ref.squares[to] = captured_piece;
        
        return in_check;
    }

    pub fn isMoveLegal(self: *MoveGenerator, move: board.Move) bool {
        _ = self;
        _ = move;
        // Simplified legal move check - would implement full validation
        return true;
    }
};